src/umpire/util/AllocationMap.cpp

```diff
@@ -14,249 +14,481 @@
+  m_tail = curr;
+  m_length++;
+}
+
+AllocationRecord RecordList::pop_back()
+{
+  AllocationRecord ret{};
+  if (m_length > 0) {
+    ret = m_tail->rec;
+    Block* prev = m_tail->prev;
+    block_pool.deallocate(m_tail);
+    m_tail = prev;
+    m_length--;
   }
+  return ret;
+}
+
+RecordListConstIterator RecordList::begin() const
+{
+  return RecordListConstIterator{this, false};
+}
+
+RecordListConstIterator RecordList::end() const
+{
+  return RecordListConstIterator{this, true};
+}
+
+size_t RecordList::size() const { return m_length; }
+bool RecordList::empty() const { return size() == 0; }
+
+AllocationRecord* RecordList::back() { return &m_tail->rec; }
 
+// RecordListConstIterator
+RecordListConstIterator::RecordListConstIterator(const RecordList* list,
+                                                 bool end)
+  : m_list(list), m_curr(end ? nullptr : m_list->m_tail)
+{
+}
+
+AllocationRecord& RecordListConstIterator::operator*() const
+{
+  // WARNING: Don't call this if at end() -- dereferences a nullptr
+  return m_curr->rec;
+}
+
+const AllocationRecord* RecordListConstIterator::operator->() const
+{
+  return m_curr ? &(m_curr->rec) : nullptr;
+}
+
+RecordListConstIterator& RecordListConstIterator::operator++()
+{
+  m_curr = m_curr->prev;
   return *this;
 }
 
-bool AllocationMap::ConstIterator::operator==(const AllocationMap::ConstIterator& other)
+RecordListConstIterator RecordListConstIterator::operator++(int)
 {
-  return (data->map == other.data->map) && (data->it == other.data->it) && (end == other.end);
+  RecordListConstIterator tmp{*this};
+  m_curr = m_curr->prev;
+  return tmp;
 }
 
-bool AllocationMap::ConstIterator::operator!=(const AllocationMap::ConstIterator& other)
+bool RecordListConstIterator::operator==(const RecordListConstIterator& other)
 {
-  return !(*this == other);
+  return m_list == other.m_list && m_curr == other.m_curr;
 }
 
-AllocationMap::ConstIterator::ConstIterator(judyL2Array<uintptr_t, uintptr_t>* map_, const bool end_) :
-  end(end_), data(new JudyL2Data{map_,
-                                 map_->begin().key,
-                                 map_->begin().value->begin()}) {}
+bool RecordListConstIterator::operator!=(const RecordListConstIterator& other)
+{
+  return !(*this == other);
+}
 
+// AllocationMap
 AllocationMap::AllocationMap() :
-  m_records(new judyL2Array<uintptr_t, uintptr_t>()),
+  m_array(nullptr),
+  m_last(nullptr),
+  m_max_levels(sizeof(uintptr_t)),
+  m_depth(1),
+  m_size(0),
   m_mutex(new std::mutex())
 {
+  // Create new judy array
+  m_array = judy_open(m_max_levels, m_depth);
 }
 
 AllocationMap::~AllocationMap()
 {
-  delete m_records;
+  // Delete all entries
+  clear();
+
+  // Close the judy array, freeing all memory.
+  judy_close(m_array);
+
+  // Delete the mutex
   delete m_mutex;
 }
 
-void
-AllocationMap::insert(void* ptr, AllocationRecord* alloc_record)
+void AllocationMap::insert(void* ptr, AllocationRecord record)
 {
   try {
```

Comment:
Will any of the code in here throw? We might be able to move the try-catch block...

Reply:
Good catch! I removed the try/catch blocks from here and findRecord.